cli/flox/src/commands/init/python.rs

```diff
@@ -0,0 +1,664 @@
+use std::borrow::Cow;
+use std::fmt::{self, Debug, Display};
+use std::path::Path;
+use std::str::FromStr;
+
+use anyhow::{anyhow, Context, Error, Result};
+use flox_rust_sdk::flox::Flox;
+use flox_rust_sdk::models::environment::path_environment::InitCustomization;
+use flox_rust_sdk::models::manifest::PackageToInstall;
+use indoc::{formatdoc, indoc};
+use itertools::Itertools;
+use log::debug;
+
+use super::{format_customization, get_default_package_if_compatible, InitHook, AUTO_SETUP_HINT};
+use crate::commands::init::get_default_package;
+use crate::utils::dialog::{Dialog, Select};
+use crate::utils::message;
+
+pub(super) struct Python {
+    providers: Vec<Provide<Box<dyn Provider>>>,
+    selected_provider: Option<Box<dyn Provider>>,
+}
+
+impl Python {
+    pub fn new(path: &Path, flox: &Flox) -> Self {
+        let providers = vec![
+            PoetryPyProject::detect(path, flox).into(),
+            PyProject::detect(path, flox).into(),
+            Requirements::detect(path, flox).into(),
+        ];
+
+        debug!("Detected Python providers: {:#?}", providers);
+
+        Self {
+            providers,
+            selected_provider: None,
+        }
+    }
+}
+
+impl InitHook for Python {
+    /// Returns `true` if any valid provider was found
+    ///
+    /// [Self::prompt_user] and [Self::get_init_customization]
+    /// are expected to be called only if this method returns `true`!
+    fn should_run(&mut self, _path: &Path) -> Result<bool> {
+        Ok(self
+            .providers
+            .iter()
+            .any(|provider| matches!(provider, Provide::Found(_))))
+    }
+
+    /// Empties the [Python::providers] and stores the selected provider in [Python::selected_provider]
+    fn prompt_user(&mut self, _path: &Path, _flox: &Flox) -> Result<bool> {
+        let mut found_providers = std::mem::take(&mut self.providers)
+            .into_iter()
+            .filter_map(|provider| match provider {
+                Provide::Found(provider) => Some(provider),
+                _ => None,
+            })
+            .collect::<Vec<_>>();
+
+        fn describe_provider(provider: &dyn Provider) -> String {
+            format!(
+                "* {} ({})\n\n{}",
+                provider.describe_provider(),
+                provider.describe_reason(),
+                textwrap::indent(&provider.describe_customization(), "  ")
+            )
+        }
+
+        message::plain(formatdoc! {"
+            Flox detected a Python project with the following Python provider(s):
```

Comment:
I agree the wording may need to be adjusted.
I think being somewhat more wordy but structured makes sense anyway whether that is in the current format or an abbreviated version (e.g. below) or something else entirely i guess we'll need input from users and @jennymahmoudi.

```
installs: ...
on-activate : ...
```

then again, at least in the python world this might not actually that uncommon to have both pyproject _and_ requirements.txts; sourcegraph finds

* [about 3k repos that have both](https://sourcegraph.com/search?q=context:global+repo:has.file%28path:requirements.txt%29+and+repo:has.file%28path:pyproject.toml%29+&patternType=keyword&sm=0&filters=%5B%5B%22type%22%2C%22Repositories%22%2C%22type%3Arepo%22%5D%5D)
* [about 9.6k repos that have either XOR the other](
https://sourcegraph.com/search?q=context:global+%28repo:has.file%28path:requirements.txt%29+and+-repo:has.file%28path:pyproject.toml%29%29+or+%28-repo:has.file%28path:requirements.txt%29+and+repo:has.file%28path:pyproject.toml%29%29&patternType=keyword&sm=0&filters=%5B%5B%22type%22%2C%22Repositories%22%2C%22type%3Arepo%22%5D%5D)

Reply:
I think `providers` is fine, and it sounds like there isn't a better or more correct word that's obvious. `Tool` is different, but not necessarily better. The headline that lists "pyproject" or "poetry" or whatever, is the key thing that someone will use to understand what we mean. 

I agree that the multiple providers structure is a little awkward for the single-provider case, and we can revisit that after we get feedback on the functionality.